added padding arg to bounding_region

shimwell opened the pull request:
# Description

Every so often a user makes a DAGMC geometry and puts the source outside of the bounding CSG cell. This has come up a few times on the forum, most recently

https://openmc.discourse.group/t/error-more-than-95-of-external-source-sites-sampled-were-rejected-please-check-your-external-source-s-spatial-definition/3343

I'm wondering if this proposed ```padding``` argument might help as it allows the newly formed bounding region to be larger and thus include the source term.

```python
u = openmc.DAGMCUniverse('dagmc.h5m').bounded_universe(padding=1000)  
# makes the containing cell bigger than the dagmc universe by 1000 in each direction
```

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have followed the [style guidelines](https://docs.openmc.org/en/latest/devguide/styleguide.html#python) for Python source files (if applicable)
- [x] I have made corresponding changes to the documentation (if applicable)
- [x] I have added tests that prove my fix is effective or that my feature works (if applicable)

makeclean commented:
I guess this kinda depends on how wrong the source is, could a user still not put the source even further away than that? Would there be a way to ask the source object its nominal position, maybe with some bounds and use those? One of the key tennets of radiation transport is know your problem, would an error message to the source sampling rejection say something like, "please ensure your source is fully contained in the non void regions of the problem?"

shimwell commented:
Yes, it would be ideal if the users know where to put their source.

I guess another problem this PR helps with is, if you have a dagmc geometry and you **want** to put the source outside of the material region then  this helps ensure you are still inside the allowable universe. As you can enlarge the vacuum surfaces to encompass the source.

pshriwise commented:
I had a thought on this yesterday. @shimwell what would you think of adding an `extend` method to the `BoundingBox` class? This would allow us to get the bounding box, call the extend method with the `padding`/`padding_distance` as the argument and then all of the logic for determining the region radius/extents can remain the same.

shimwell commented:
Thanks all, I have had a go at addressing your suggestions

- I've added ```extend``` to the ```BoundingBox``` class and moved the padding logic in there.

- I've corrected the radius equation as spotted by Ethan

- I've changed the argument name to ```padding_distance``` 

Unfortunately, I've not managed to remove the need for ```starting_id``` but agree it would be nice if it was not needed.

pshriwise commented on openmc/bounding_box.py:
```diff
@@ -87,3 +87,24 @@ def extent(self):
+            The distance to enlarge the bounding box by
+        """
+
+        return BoundingBox(np.array(
+            [
+                self[0][0] - padding_distance,
+                self[0][1] - padding_distance, 
+                self[0][2] - padding_distance
+            ]
+        ), np.array(
+            [
+                self[1][0] + padding_distance,
+                self[1][1] + padding_distance, 
+                self[1][2] + padding_distance
+            ]
+        ))
```
```suggestion
        self[0] -= padding_distance
        self[1] += padding_distance
```

I stared at this a bit to figure out why the suggestion above wouldn't work -- because this class inherits from tuple. I'm feeling we should inherit from a mutable type for this object. From working with bounding volume hierarchies, mutable boxes allow for really convenient operations like `bbox.update(other_bbox)` or `bbox.update(point)` to make sure the box contains another box or other points, respectively. With an immutable box, we end up creating potentially more objects than we need b/c we can't modify the underlying data.

If you agree @shimwell, this can definitely be handled in another PR and we can leave this method as-is for now. Just running it by you and @eepeterson to see what you think.

eepeterson replied:
I agree with the benefits of mutability here, but don't see any clear need for inheriting from anything in particular in this case. I think we could get away just having a private `_data` attribute that houses the `2x3 np.ndarray` that represents the bounding box.

shimwell replied:
I would be keen on anything that allows us to check if a point is inside the bounding box

pshriwise replied:
Not quite what that method would do, but we can add that in! Let's tackle that separately from this capability though. Thank you for this update @shimwell! I think it cleans up the padding process a bit.

pshriwise reviewed (approved):
Thanks for this addition @shimwell!
